sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -158,25 +158,34 @@ case class AdaptiveSparkPlanExec(
       var result = createQueryStages(currentPhysicalPlan)
       val events = new LinkedBlockingQueue[StageMaterializationEvent]()
       val errors = new mutable.ArrayBuffer[SparkException]()
+      val runningStages = util.Collections.synchronizedSet(new util.HashSet[QueryStageExec]())
       var stagesToReplace = Seq.empty[QueryStageExec]
       while (!result.allChildStagesMaterialized) {
         currentPhysicalPlan = result.newPlan
         if (result.newStages.nonEmpty) {
           stagesToReplace = result.newStages ++ stagesToReplace
           executionId.foreach(onUpdatePlan)
 
-          // Start materialization of all new stages.
-          result.newStages.foreach { stage =>
+          // Start materialization of all new stages and fail fast if any stages failed eagerly
+          var anyFail = false
+          val stages = result.newStages.toIterator
+          while (!anyFail && stages.hasNext) {
+            val stage = stages.next()
+            runningStages.add(stage)
             try {
               stage.materialize().onComplete { res =>
+                runningStages.remove(stage)
                 if (res.isSuccess) {
                   events.offer(StageSuccess(stage, res.get))
                 } else {
                   events.offer(StageFailure(stage, res.failed.get))
                 }
               }(AdaptiveSparkPlanExec.executionContext)
             } catch {
-              case e: Throwable => events.offer(StageFailure(stage, e))
+              case e: Throwable =>
+                anyFail = true
+                runningStages.remove(stage)
```

Comment:
This is the one we need to track, as `QueryStageExec` can't tell if it fails at the very beginning.

Can we only track one stage which fails at the beginning?